src/app/shared/services/dynamic-data/dynamic-data.service.ts

```diff
@@ -0,0 +1,196 @@
+import { Injectable } from "@angular/core";
+import { addRxPlugin, MangoQuery, RxDocument } from "rxdb";
+import { firstValueFrom, map } from "rxjs";
+
+import { FlowTypes } from "packages/data-models";
+import { environment } from "src/environments/environment";
+import { AppDataService } from "../data/app-data.service";
+import { AsyncServiceBase } from "../asyncService.base";
+import { arrayToHashmap, deepMergeObjects } from "../../utils";
+import { PersistedMemoryAdapter } from "./adapters/persistedMemory";
+import { ReactiveMemoryAdapater, REACTIVE_SCHEMA_BASE } from "./adapters/reactiveMemory";
+import { TemplateActionRegistry } from "../../components/template/services/instance/template-action.registry";
+import { TopLevelProperty } from "rxdb/dist/types/types";
+
+type IDocWithID = { id: string };
+
+@Injectable({ providedIn: "root" })
+/**
+ * The dynamic data service handles the process of loading data from sheets, storing
+ * in memory as a reactive data object, and enabling a persisted write-layer on top
+ * for storing user changes
+ */
+export class DynamicDataService extends AsyncServiceBase {
+  /**
+   * The main DB stores in-memory the rows of any requested sheets, merged with user writes
+   *
+   * Each flow is represented in its own collection, and populated as requested.
+   * This allows users to query and subscribe to data changes in an efficient way
```

Comment:
I'm just learning about NoSQL (I'm currently working on an app using Firestore). I've been surprised by the fact that you can't query across multiple collections. Storing each flow in its own collection seems the neatest approach here, and I know it is always possible to combine data from multiple reads if we ever wanted to query multiple data_lists

Reply:
Firestore does expose a method for [collectionGroup queries](https://firebase.google.com/docs/firestore/query-data/queries#collection-group-query), however usually you need to define a custom index in advance and support is somewhat limited (mostly cases where multiple collections have subcollections or documents with the same name)

The nice thing about using an in-memory database is that if we later decide we want to change how it is used we don't have schema-breaking changes to worry about (because nothing is persisted). It's also for that reason that the caching layer actually stores all documents in a single table, indexed by flow_type and flow_name, and a single `data` column that stores the full json of any changes to persist onto that flow